### ISSUE TITLE
Angle-instanced-arrays-out-of-bounds.html contains ambigous error condition check

### ISSUE THREAD

**Issue by RafaelCintron** [KhronosGroup/WebGL#2500]
The WebGL 1.0 test [angle-instanced-arrays-out-of-bounds.html ](https://www.khronos.org/registry/webgl/sdk/tests/conformance/extensions/angle-instanced-arrays-out-of-bounds.html?webglVersion=1&quiet=0) contains the following section:

````
Test with an empty buffer for the instanced attribute
...
PASS getError was expected value: NO_ERROR : after evaluating: ext.drawElementsInstancedANGLE(gl.TRIANGLES, 10000, gl.UNSIGNED_BYTE, 0, 0)
...
````

Note that we pass a primCount of zero but a count of 10000.  This call to drawElementsInstancedANGLE is technically incorrect since the element array buffer only has 9 items and we are passing a count of 10000.  However, many implementations will no-op the call since the primCount is 0 and not bother to check for other errors.  ANGLE, on the other hand, performs the no-op check after checking for element array buffer out of bounds. Hence, the test fails.  

````C++
bool ValidateDrawElementsInstancedBase(ValidationContext *context,
                                       GLenum mode,
                                       GLsizei count,
                                       GLenum type,
                                       const GLvoid *indices,
                                       GLsizei primcount)
{
    if (primcount < 0)
    {
        context->handleError(InvalidValue() << "primcount cannot be negative.");
        return false;
    }

    if (!ValidateDrawElementsCommon(context, mode, count, type, indices, primcount))
    {
        return false;
    }

    // No-op zero primitive count
    return (primcount > 0);
}
````

What is the right answer here?  Should early outs win or should error checking win?  I'm inclined to go with the latter for this particular case. If you call drawElementsInstanced and pass a primitive count of zero, you're doing it wrong.

**Comment by greggman** [comment]
I'm not sure I'm conviced

it is perfectly reasonable to be able to pass a primcount of 0.

let's say you have an interactive demo with a slider that goes from 0 to
1000 primcount. You shouldn't have to special case zero.

whether or not count can be larger then your buffers when primcount is zero
is a separate issue from allowing primcount to be zero.

erroring when count is too large and primcount is zero seems to be
inconsistent with other parts of the spec. Youre not asking to draw so
nothing is out of bounds.

I also dont see how it's technically incorrect
real GL doesn't check so it would run just fine

**Comment by kainino0x** [comment]
> Youre not asking to draw so nothing is out of bounds.

Good points. I think this is a convincing argument in favor of letting the early-out win.

**Comment by RafaelCintron** [comment]
@greggman, you make a compelling argument. 

On the flipside, there is an argument to be made that browsers, runtimes, and drivers should be as thin as possible and not contain "helpful" if statements to avoid unnecessary calls.   If you've created an optimized game engine that culls as much of the scene as possible, only drawing objects that are absolutely necessary, the "helpful" if statements at multiple levels of the stack only make things slower for you.  

From my reading of the specs, there is nothing that requires error checking to be skipped if the primCount or the count == 0.  Please correct me if I am wrong.



On a related note, I've recently encountered a similar mismatch when it comes to framebufferTexture2D.  (Haven't had a chance to file an issue yet) The OpenGL ES spec says that passing 0 for the texture argument should ignore the texTarget and level arguments.  Edge obeys this clause of the spec but ANGLE does the error checking regardless of what you pass for the texture argument.  This clause makes sense since nulling out an attachment is a legit thing to do.

**Comment by greggman** [comment]
> From my reading of the specs, there is nothing that requires error
checking to be skipped if the primCount or the count == 0. Please correct
me if I am wrong

which spec are we talking about ? maybe I missing it but I thought opengl
es did not do bounds checks so there is no error for primcount = 0, count =
10000, buffer too small.

the only check is a WebGL check . for other areas of the spec webgl only
requires you can't read out of bounds data. since a primcount of 0 = not
reading of bounds data it seems consistent not to error to me. I'm happy to
be corrected.

**Comment by RafaelCintron** [comment]
@kenrussell and @jdashg, what are your thoughts?

**Comment by jdashg** [comment]
Ah, so I've actually just been thinking about this! (I just updated our vertex fetch validation in Gecko)

We should do all ordinary error checking for `!primCount||!vertCount`, but we do not need to generate the errors we would otherwise generate for out-of-bounds vertex fetching. (which is really not detailed in the GL/GLES specs)
For instance, even if `DrawArrays(POINTS, 0, 1)` doesn't have enough data in its active VBOs, `DrawArraysInstanced(POINTS, 0, N, 0)` and `DrawArraysInstanced(POINTS, 0, 0, N)` could still succeed. (NB: All `DrawArrays(...)` calls are actually `DrawArraysInstanced(..., 1)`)

`DrawArrays(X, 0, 0)` could still fail for any check unrelated to the /active vertex attribs/, such as the check that all (even non-active) enabled vertex attrib arrays have a non-null buffer binding. (though the size of the bound buffer could be 0!)
Texture feedback tests might also be omittable, but I don't love that.

`DrawArrays(X, 0, 0)` should definitely fail for, for example, framebuffer incompleteness.

I think that DrawArrays(X, 0, 0) should generally fail for all the same reasons that DrawArrays(X, 0, 1) should fail for, since we really shouldn't optimize or special case the uncommon case here. In Gecko, we do every normal draw test, but the active vertex fetch tests all pass pretty trivially. (fetching 0 of something is never unsafe, even if it may arguably be out-of-bounds, such as fetching 0 bytes at offset=1 from buffer.size=0)

I really don't want to mandate early-out, since I really don't want to add more branches or complexity to draw validation code.

TL;DR:
We should do all normal checks for no-fetch draw calls.
(Or doing such checks should be allowed by the spec)

**Comment by RafaelCintron** [comment]
@jdashg , @kainino0x , @greggman , @kenrussell , have we come to an agreement on this issue?  I'm inclined to favor Jeff's suggestion of doing normal checks for no-fetch draw calls.  That would mean changing this test to flag INVALID_OPERATION instead of allowing NO_ERROR like it does today.

**Comment by kainino0x** [comment]
I have no objections.

**Comment by kenrussell** [comment]
Sorry for the delay. Sounds OK to me to do the normal checks for zero-length draw calls.

**Comment by jdashg** [comment]
I'm not sure the status of this test should change. This passes in Firefox, and I believe we still do non-fetch validation on non-fetch draw calls. We don't do partial-fetch validation (validate indexed fetches if primcount=0). I would be fine with allowing this validation to happen, though I'm slightly against requiring anything in particular here, as long as we're generating errors for invalid fetches.

Specifically, the test in question would only be an invalid fetch if any vertices were fetched, which, for vertCount=0||primCount=0, none are. (since the index lookup should never even be done, it shouldn't matter if the index buffer is too small) At least in Firefox, since we don't consider any fetch to happen in this case, we don't validate index buffer values.

**Comment by RafaelCintron** [comment]
@jdashg , I am confused.  Earlier in the issue you stated "I really don't want to mandate early-out, since I really don't want to add more branches or complexity to draw validation code. ... We should do all normal checks for no-fetch draw calls."

In your latest reply you state: "Specifically, the test in question would only be an invalid fetch if any vertices were fetched, which, for vertCount=0||primCount=0, none are. At least in Firefox, since we don't consider any fetch to happen in this case, we don't validate index buffer length or values."

Are you having second thoughts or am I misunderstanding your replies?

**Comment by jdashg** [comment]
Indexed draw calls with no verts or no prims never actually fetch from the
index buffer, therefore so long as the index buffer is non-null (which we
mandate explicitly elsewhere), even a 0-sized index buffer (and indeed
0-sized vertex buffers) is /technically/ large enough for any
no-vert/no-prim inputs to a draw call. In Firefox, we validate this as a
0-size fetch, which is trivially valid. We are doing all normal checks,
despite the fetch validation being trivial. We do not skip subsequent
validation after realizing it's a 0-size fetch, which at least some
browsers have done in the past.

**Comment by Oletus** [comment]
Where is the spec text that actually specifies the error to be generated on too large count passed to drawElements? There's the "Enabled Vertex Attributes and Range Checking" section in the WebGL 1.0 spec, but that only clearly talks about out of range indices, not out of range access to the ELEMENT_ARRAY_BUFFER. Maybe that could use a clarification?

I'm in favor of not generating an error if the draw call is otherwise valid, instance count is 0 and count would cause indices outside the ELEMENT_ARRAY_BUFFER to be referenced in case instance count were > 0.

**Comment by jdashg** [comment]
So IMO what we should do here is do the error checking for general draw calls (current_program != null), but only validate bounds checking for the actual requested vertex/index buffer fetches. (which would be trivially in-bounds for vertexCount==0 || instanceCount==0) A fetch of vertexCount:10000/instanceCount:0 fetches no vertex data, and is therefore not an out-of-bounds INVALID_OP. (RBAB means this should be allowed to not generate an error regardless, though!)

I do agree that we don't seem to have actual wording specifying that INVALID_OP might occur when requesting an out-of-bounds subrange of the index-buffer.

**Comment by kenrussell** [comment]
Fixed by @jdashg in #2607 .
